### ISSUE TITLE
Consider more generic ways to invert channels

### ISSUE THREAD

**Issue by flying-sheep** [bjz/color-rs#24]
L*a*b colors are usually represented as one of the following two

L* | a* | b*
--- | --- | ---
0 – 100 | -120 – 120 |  -120 – 120
0 – 100 | -128 – 128 |  -128 – 128

the last one more “accomodating” to computers (to *almost* represent them as `i8`, i guess).

now adding this is a problem: we’d have to add two more ways to invert this: `100 - L` and the simple `-a`/`-b`.

i think we should consider a way to handle this, e.g. by tying the kind of channel to the color model:

 | HSV | L*a*b* | RGB
--- | --- | --- | ---
1 | circular | unsigned | unsigned
2 | unsigned | signed | unsigned
3 | unsigned | signed | unsigned

that would involve either normalizing float channels (always clamp to [-1..1] or [0..1], modulo when circular). also bits would be lost: using 3×i8 for L*a*b* would mean using all int values from -128 to 127, but only 100 lightness levels!

---

do you have a different idea how to handle this?

**Comment by flying-sheep** [comment]
yeah, sure. But if that would be meaningful, we’d need `l:u8, a:i8, b:i8`, which isn’t supported by the current generics model (one type of numeric primitive as type parameter)

**Comment by retep998** [comment]
Perhaps we could have tag types for the 4 sizes of integers, with a trait + associated types specifying the signed and unsigned versions. Thus you can specialize the color models on the tag types and have signed ints for some components and unsigned ints for other components.

**Comment by flying-sheep** [comment]
sorry, i’m a rust noob: tag types?

**Comment by retep998** [comment]
Something like thus
```rust
struct Int8;
impl Something for Int8 {
    type Signed = i8;
    type Unsigned = u8;
}
```

**Comment by flying-sheep** [comment]
ah, that brand new thing! very good idea.

**Comment by bjz** [comment]
Thanks for spending the time to write this up! I will have to spend some digesting this. Are you available on #rust-gamedev at any time to have a chat?

**Comment by flying-sheep** [comment]
sure, i’ll hop on right now!
